src/google-classroom-extractor/google_classroom_extractor/mapping/user_submission_activities.py

```diff
@@ -0,0 +1,154 @@
+    SOURCE_SYSTEM,
+    ENTITY_STATUS_ACTIVE,
+)
+
+
+def submissions_to_user_submission_activities_dfs(
+    submissions_df: DataFrame,
+) -> Dict[str, DataFrame]:
+    """
+    Convert a Submission API DataFrame to a Dict of UserActivity
+    UDM DataFrames grouped by source system section id.
+
+    Parameters
+    ----------
+    submissions_df: DataFrame
+        is a Submission API DataFrame
+
+    Returns
+    -------
+    Dict[str, DataFrame] LMS UDM UserActivity DataFrames
+        grouped by source system section id
+
+    Notes
+    -----
+    UserActivity DataFrame columns are:
+        ActivityDateTime: The date/time the activity occurred
+        ActivityStatus: The activity status
+        ActivityTimeInMinutes: The total activity time in minutes
+        ActivityType: The type of activity, here "Submission" or "Grade"
+        AssignmentIdentifier: A unique numeric identifier assigned to the assignment
+        Content: Content associated with the activity
+        EntityStatus: The status of the record
+        LMSSectionIdentifier: A unique numeric identifier assigned to the section
+        SourceSystem: The system code or name providing the user data
+        SourceSystemIdentifier: A unique number or alphanumeric code assigned to a
+            user by the source system
+        LMSUserIdentifier: A unique numeric identifier assigned to the user
+        LMSUserActivityIdentifier: A unique numeric identifier assigned to the user activity
+        CreateDate: Date this record was created
+        LastModifiedDate: Date this record was last updated
+    """
+    assert isinstance(submissions_df, DataFrame)
+    assert "submissionHistory" in submissions_df.columns
+    assert "id" in submissions_df.columns
+    assert "courseId" in submissions_df.columns
+    assert "courseWorkId" in submissions_df.columns
+
+    # convert json-like submissionHistory string to list of dicts
+    submissions_df["submissionHistory"] = submissions_df["submissionHistory"].apply(lambda json_like: json.loads(json_like.replace("'", '"')))
+
+    submissions_df = submissions_df[["id", "courseId", "courseWorkId", "submissionHistory"]]
+
+    # explode submissionHistory lists into rows with other columns duplicated
+    history_df = submissions_df.explode("submissionHistory")
+
+    # expand submissionHistory dicts (stateHistory and gradeHistory) into their own columns
+    history_df = history_df["submissionHistory"].apply(Series).merge(history_df, left_index=True, right_index=True, how='outer')
+    history_df.drop(columns=["submissionHistory"], inplace=True)
+
+    # expand stateHistory
+    state_history_df = concat([history_df, history_df["stateHistory"].apply(Series)], axis=1)
+    state_history_df.dropna(subset=["stateHistory"], inplace=True)
+
+    # enrich state_history_df
+    state_history_df["SourceSystemIdentifier"] = "S-" + state_history_df[
+        ["courseId", "courseWorkId", "id", "stateTimestamp"]
+    ].agg("-".join, axis=1)
```

Comment:
In Schoology attendance events there is no source system identifier, so I had to create one by concatenating the enrollment id and the event date, giving me something like `123456-2020-10-30`. Then I found it a little odd for dashes to have two different meanings, and switched to `123456#2020-10-30`. Ultimately internal self-consistency in the application is all that matters. Thoughts?

Reply:
Yeah, this was the first time I had to include timestamps to differentiate. Hadn't thought about the double use of dashes. Not too concerned but open to changing. Hash is a good choice if we want to change as submission ids in GC can have underscores.

Reply:
Leaving this to your judgment for now. Maybe a good thing to ask for feedback on from end-users ... assuming we ever have end users 😆 . 

![I believe in the users](https://fastly.syfy.com/sites/syfy/files/styles/1200x680/public/wire/legacy/Tron-Light-Cycle.jpg?offset-y=0)